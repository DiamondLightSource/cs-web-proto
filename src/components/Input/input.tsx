import React, { useState } from "react";
import { connectionWrapper } from "../ConnectionWrapper/connectionWrapper";
import { writePv } from "../../hooks/useCs";
import { VType } from "../../vtypes/vtypes";

import classes from "./input.module.css";
import { vtypeToString, stringToVtype } from "../../vtypes/utils";

export interface InputProps {
  pvName: string;
  value: string;
  onKeyDown: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onBlur: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onClick: (event: React.MouseEvent<HTMLInputElement>) => void;
  style?: object;
}

export const Input: React.FC<InputProps> = (props: InputProps): JSX.Element => (
  <input
    type="text"
    value={props.value}
    onKeyDown={props.onKeyDown}
    onChange={props.onChange}
    onBlur={props.onBlur}
    onClick={props.onClick}
    className={`Input ${classes.Input}`}
    style={props.style}
  />
);

interface ConnectedInputProps {
  pvName: string;
  style?: object;
}

interface SmartInputProps {
  pvName: string;
  value?: VType;
  style?: object;
}

export const SmartInput: React.FC<SmartInputProps> = (
  props: SmartInputProps
): JSX.Element => {
  const [inputValue, setInputValue] = useState("");
  const [editing, setEditing] = useState(false);
  function onKeyDown(event: React.KeyboardEvent<HTMLInputElement>): void {
    if (event.key === "Enter") {
      writePv(props.pvName, stringToVtype(event.currentTarget.value));
      setInputValue("");
      setEditing(false);
    }
  }
  function onChange(event: React.ChangeEvent<HTMLInputElement>): void {
    setInputValue(event.currentTarget.value);
  }
  function onClick(event: React.MouseEvent<HTMLInputElement>): void {
    /* When focus gained allow editing. */
    if (!editing) {
      setInputValue("");
      setEditing(true);
    }
  }
  function onBlur(event: React.ChangeEvent<HTMLInputElement>): void {
    setEditing(false);
    /* When focus lost show PV value. */
    setInputValue(vtypeToString(props.value));
  }

  if (!editing && inputValue !== vtypeToString(props.value)) {
    setInputValue(vtypeToString(props.value));
  }

  return (
    <Input
      pvName={props.pvName}
      value={inputValue}
      onKeyDown={onKeyDown}
      onChange={onChange}
      onBlur={onBlur}
      onClick={onClick}
      style={props.style}
    />
  );
};

export const ConnectedInput: React.FC<ConnectedInputProps> = connectionWrapper(
  SmartInput
);
