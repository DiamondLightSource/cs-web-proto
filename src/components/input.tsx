import React, { useState } from "react";
import { useSelector } from "react-redux";
import { useSubscription, writePv } from "../hooks/useCs";
import { CsState } from "../redux/csState";

export interface InputProps {
  value: string;
  onKeyDown: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onBlur: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onClick: (event: React.MouseEvent<HTMLInputElement>) => void;
}

export const Input: React.FC<InputProps> = (props: InputProps): JSX.Element => (
  <input
    type="text"
    value={props.value}
    onKeyDown={props.onKeyDown}
    onChange={props.onChange}
    onBlur={props.onBlur}
    onClick={props.onClick}
  />
);

interface ConnectedInputProps {
  pvName: string;
}

export const ConnectedInput: React.FC<ConnectedInputProps> = (
  props: ConnectedInputProps
): JSX.Element => {
  const [inputValue, setInputValue] = useState("");
  const [editing, setEditing] = useState(false);
  useSubscription(props.pvName);
  /* It would be nice to be able to share a selector, but it's
     not immediately obvious how to make a selector that takes
     an argument other than state. */
  const latestValue: string = useSelector((state: CsState): string => {
    let value = state.valueCache[props.pvName];
    if (value == null) {
      return "";
    } else {
      return value.value.toString();
    }
  });
  function onKeyDown(event: React.KeyboardEvent<HTMLInputElement>): void {
    if (event.key === "Enter") {
      writePv(props.pvName, {
        type: "NTScalar",
        value: event.currentTarget.value
      });
      setInputValue("");
    }
  }
  function onChange(event: React.ChangeEvent<HTMLInputElement>): void {
    setInputValue(event.currentTarget.value);
  }
  function onClick(event: React.MouseEvent<HTMLInputElement>): void {
    /* When focus gained allow editing. */
    if (!editing) {
      setInputValue("");
      setEditing(true);
    }
  }
  function onBlur(event: React.ChangeEvent<HTMLInputElement>): void {
    setEditing(false);
    /* When focus lost show PV value. */
    setInputValue(latestValue);
  }

  if (!editing && inputValue !== latestValue) {
    setInputValue(latestValue);
  }

  return (
    <Input
      value={inputValue}
      onKeyDown={onKeyDown}
      onChange={onChange}
      onBlur={onBlur}
      onClick={onClick}
    />
  );
};
