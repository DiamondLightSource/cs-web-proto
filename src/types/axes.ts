import { Color } from "./color";
import { Font, FontStyle } from "./font";

export class Axis {
  public index: number;
  public autoScale?: boolean;
  public autoScaleThreshold?: number;
  public axisColor?: Color;
  public axisTitle?: string;
  public showGrid?: boolean;
  public gridColor?: Color;
  public dashGridLine?: boolean;
  public timeFormat?: number;
  public scaleFormat?: string;
  public scaleFont?: Font;
  public titleFont?: Font;
  public visible?: boolean;
  public logScale?: boolean;
  public leftBottomSide?: boolean;
  public maximum?: number;
  public minimum?: number;
  public yAxis?: boolean;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property. Uses same
   * default values as csstudio.opibuilder.xygraph.
   */
  public constructor(idx: number) {
    this.index = idx ?? 0;
    this.autoScale = this.autoScale ?? true;
    this.autoScaleThreshold = this.autoScaleThreshold ?? 0.95;
    this.axisColor = this.axisColor ?? new Color("rgb(0, 0, 0");
    this.axisTitle = this.axisTitle ?? "";
    this.showGrid = this.showGrid ?? false;
    this.gridColor = this.gridColor ?? new Color("rgb(0, 0, 0");
    this.dashGridLine = this.dashGridLine ?? true;
    this.timeFormat = this.timeFormat ?? 0;
    this.scaleFormat = this.scaleFormat ?? "";
    this.scaleFont = this.scaleFont ?? new Font();
    this.titleFont = this.titleFont ?? new Font(10, FontStyle.Bold);
    this.visible = this.visible ?? true;
    this.logScale = this.logScale ?? false;
    this.leftBottomSide = this.leftBottomSide ?? true;
    this.maximum = this.maximum ?? 0;
    this.minimum = this.minimum ?? 100;
    // If axis has index 1 or above, assume y axis
    let axisBool = false;
    if (this.index > 0) {
      axisBool = true;
    }
    this.yAxis = this.yAxis ?? axisBool;
  }
}

export class Axes {
  public count: number;
  public axisOptions: Axis[];

  public constructor(count: number, axes: Axis[]) {
    if (count !== axes.length) {
      throw new Error(
        `Count ${count} is not equal to number of axes ${axes.length}`
      );
    }
    this.count = count;
    this.axisOptions = axes;
  }
}
