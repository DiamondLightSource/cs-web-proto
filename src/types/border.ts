import { Color } from "./color";

export enum BorderStyle {
  None,
  Line,
  Dashed,
  Dotted,
  GroupBox
}

const CssBorders: { [key in BorderStyle]: string } = {
  [BorderStyle.None]: "none",
  [BorderStyle.Line]: "solid",
  [BorderStyle.Dashed]: "dashed",
  [BorderStyle.Dotted]: "dotted",
  [BorderStyle.GroupBox]: "ridge"
};

export class Border {
  private style: BorderStyle;
  private color: Color;
  // This will render a border width in pixels. I expect that we will revisit this
  // at some later point, possibly to allow or fix to rems.
  private width: number;

  public constructor(style: BorderStyle, color: Color, width: number) {
    this.style = style;
    this.color = color;
    this.width = width;
  }

  public css(): object {
    return {
      borderStyle: CssBorders[this.style],
      borderWidth: `${this.width}px`,
      borderColor: this.color.rgbaString()
    };
  }
}
