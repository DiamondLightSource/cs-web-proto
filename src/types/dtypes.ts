import { Alarm, ALARM_NONE } from "./alarm";
import { Display, DISPLAY_NONE } from "./display";

export class DTime {
  public datetime: Date;

  public constructor(datetime: Date) {
    this.datetime = datetime;
  }
}

export function dtimeNow(): DTime {
  return new DTime(new Date());
}

interface DTypeValue {
  stringValue?: string;
  doubleValue?: number;
  arrayValue?: number[];
}

export class DType {
  public value: DTypeValue;
  public time: DTime;
  public alarm?: Alarm;
  public display?: Display;

  public constructor(
    value: DTypeValue,
    alarm?: Alarm,
    time?: DTime,
    display?: Display
  ) {
    // TODO check for no value.
    this.value = value;
    this.alarm = alarm;
    this.time = time ?? dtimeNow();
    this.display = display;
  }

  public getStringValue(): string {
    if (this.value.stringValue) {
      return this.value.stringValue;
    } else if (this.value.doubleValue) {
      return this.value.doubleValue?.toString();
    } else if (this.value.arrayValue) {
      return this.value.arrayValue?.toString();
    } else {
      return "";
    }
  }

  public getDoubleValue(): number {
    // TODO what if not defined?
    if (this.value.doubleValue) {
      return this.value.doubleValue;
    } else {
      return NaN;
    }
  }

  public getAlarm(): Alarm {
    return this.alarm ?? ALARM_NONE;
  }
  public getTime(): DTime {
    return this.time;
  }
  public getDisplay(): Display | undefined {
    return this.display ?? DISPLAY_NONE;
  }

  public toString(): string {
    return this.getStringValue();
  }
}

export function dtypeToString(dtype?: DType): string {
  if (dtype) {
    return dtype.getStringValue();
  } else {
    return "";
  }
}

export function valueToDtype(
  value: any,
  alarm = ALARM_NONE,
  time = dtimeNow(),
  display = DISPLAY_NONE
): DType {
  const dvalue: DTypeValue = {};
  if (typeof value === "string") {
    dvalue.stringValue = value;
  } else if (typeof value === "number") {
    dvalue.doubleValue = value;
  } else {
    dvalue.doubleValue = 0;
  }
  return new DType(dvalue, alarm, time, display);
}
