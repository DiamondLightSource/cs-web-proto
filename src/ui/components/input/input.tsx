import React, { CSSProperties, useState } from "react";

export const InputComponent = (props: {
  value: string;
  onEnter: (value: string) => void;
  readonly?: boolean;
  style?: CSSProperties;
  className?: string;
}): JSX.Element => {
  const [inputValue, setInputValue] = useState("");
  const [editing, setEditing] = useState(false);
  function onKeyDown(event: React.KeyboardEvent<HTMLInputElement>): void {
    if (event.key === "Enter") {
      props.onEnter(event.currentTarget.value);
      setInputValue("");
      setEditing(false);
      event.currentTarget.blur();
    }
  }
  function onChange(event: React.ChangeEvent<HTMLInputElement>): void {
    setInputValue(event.currentTarget.value);
  }
  function onClick(event: React.MouseEvent<HTMLInputElement>): void {
    /* When focus gained allow editing. */
    if (!props.readonly && !editing) {
      setInputValue("");
      setEditing(true);
    }
  }
  function onBlur(event: React.ChangeEvent<HTMLInputElement>): void {
    setEditing(false);
  }

  if (!editing && inputValue !== props.value) {
    setInputValue(props.value);
  }

  return (
    <input
      type="text"
      value={inputValue}
      readOnly={props.readonly}
      onKeyDown={onKeyDown}
      onChange={onChange}
      onBlur={onBlur}
      onClick={onClick}
      className={props.className}
      style={props.style}
    />
  );
};
