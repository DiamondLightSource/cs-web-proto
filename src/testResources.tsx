import React from "react";
import { DType, DAlarm } from "./types/dtypes";
import { FileProvider, PageState, TabState } from "./fileContext";
import { render, RenderResult } from "@testing-library/react";
import { Provider } from "react-redux";
import { MacroContext } from "./types/macros";
import { csReducer, CsState } from "./redux/csState";
import { createStore } from "redux";
import { BrowserRouter as Router } from "react-router-dom";
import {
  OPEN_WEBPAGE,
  WidgetAction,
  WritePv,
  WRITE_PV
} from "./ui/widgets/widgetActions";

// Helper functions for dtypes.
export function ddouble(
  doubleValue: number,
  alarm: DAlarm = DAlarm.NONE
): DType {
  return new DType({ doubleValue: doubleValue }, alarm);
}

export function ddoubleArray(
  arrayValue: number[],
  alarm: DAlarm = DAlarm.NONE
): DType {
  return new DType({ arrayValue: Float64Array.from(arrayValue) }, alarm);
}

export function dstring(
  stringValue: string,
  alarm: DAlarm = DAlarm.NONE
): DType {
  return new DType({ stringValue: stringValue }, alarm);
}

// Test actions
export const WRITE_PV_ACTION: WritePv = {
  type: WRITE_PV,
  writePvInfo: {
    pvName: "PV",
    value: "value",
    description: "write value to PV"
  }
};

export const WRITE_PV_ACTION_NO_DESC: WritePv = {
  type: WRITE_PV,
  writePvInfo: {
    pvName: "PV",
    value: "value"
  }
};

export const OPEN_BBC_ACTION: WidgetAction = {
  type: OPEN_WEBPAGE,
  openWebpageInfo: { url: "https://bbc.co.uk", description: "BBC" }
};

export const ACTIONS_EX_AS_ONE = {
  actions: [WRITE_PV_ACTION, WRITE_PV_ACTION_NO_DESC],
  executeAsOne: true
};

export const ACTIONS_EX_FIRST = {
  actions: [WRITE_PV_ACTION, WRITE_PV_ACTION_NO_DESC],
  executeAsOne: false
};

// Helper function for rendering with a working fileContext.
export function contextRender(
  component: JSX.Element,
  initialPageState: PageState = {},
  initialTabState: TabState = {},
  initialCsState: CsState = {
    effectivePvNameMap: {},
    globalMacros: {},
    subscriptions: {},
    valueCache: {},
    deviceCache: {}
  }
): RenderResult {
  const ParentComponent = (props: { child: JSX.Element }): JSX.Element => {
    // Hard-code macros for now.
    // eslint-disable-next-line no-template-curly-in-string
    const contextMacros = { a: "A", b: "B", c: "C", e: "${a}" };
    const globalMacros = { c: "D", d: "E" };
    initialCsState.globalMacros = globalMacros;
    const macroContext = { macros: contextMacros, updateMacro: (): void => {} };
    const store = createStore<CsState, any, any, any>(
      csReducer,
      initialCsState
    );
    return (
      <Router>
        <Provider store={store}>
          <MacroContext.Provider value={macroContext}>
            <FileProvider
              initialPageState={initialPageState}
              initialTabState={initialTabState}
            >
              {props.child}
            </FileProvider>
          </MacroContext.Provider>
        </Provider>
      </Router>
    );
  };
  return render(<ParentComponent child={component}></ParentComponent>);
}
