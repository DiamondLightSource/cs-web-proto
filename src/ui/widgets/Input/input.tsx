import React, { useState } from "react";

import classes from "./input.module.css";
import { writePv } from "../../hooks/useSubscription";
import { commonCss, Widget } from "../widget";
import { PVInputComponent, PVWidgetPropType } from "../widgetProps";
import { registerWidget } from "../register";
import {
  InferWidgetProps,
  FontPropOpt,
  ChoicePropOpt,
  ColorPropOpt,
  BoolPropOpt
} from "../propTypes";
import { Font } from "../../../types/font";
import { Color } from "../../../types/color";
import { AlarmQuality, DType } from "../../../types/dtypes";

export interface InputProps {
  pvName: string;
  value: string;
  readonly: boolean;
  foregroundColor?: Color;
  backgroundColor?: Color;
  transparent: boolean;
  alarm: AlarmQuality;
  alarmSensitive: boolean;
  onKeyDown: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onBlur: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onClick: (event: React.MouseEvent<HTMLInputElement>) => void;
  font?: Font;
  textAlign?: "left" | "center" | "right";
}

export const InputComponent: React.FC<InputProps> = (
  props: InputProps
): JSX.Element => {
  let allClasses = classes.Input;
  const style = commonCss(props);
  if (props.textAlign) {
    style.textAlign = props.textAlign;
  }
  style.color = props.foregroundColor?.toString();
  style.backgroundColor = props.backgroundColor?.toString();
  // Transparent prop overrides backgroundColor.
  if (props.transparent) {
    style["backgroundColor"] = "transparent";
  }
  if (props.readonly) {
    allClasses += ` ${classes.readonly}`;
  }
  if (props.alarmSensitive) {
    switch (props.alarm) {
      case AlarmQuality.UNDEFINED:
      case AlarmQuality.INVALID:
      case AlarmQuality.CHANGING:
        style.color = "var(--invalid)";
        break;
      case AlarmQuality.WARNING:
        style.color = "var(--alarm)";
        break;
      case AlarmQuality.ALARM:
        style.color = "var(--alarm)";
        break;
    }
  }
  return (
    <input
      type="text"
      value={props.value}
      onKeyDown={props.onKeyDown}
      onChange={props.onChange}
      onBlur={props.onBlur}
      onClick={props.onClick}
      className={allClasses}
      style={style}
      readOnly={props.readonly}
    />
  );
};

export const SmartInputComponent = (
  props: PVInputComponent & {
    font?: Font;
    foregroundColor?: Color;
    backgroundColor?: Color;
    transparent?: boolean;
    alarmSensitive?: boolean;
    textAlign?: "left" | "center" | "right";
  }
): JSX.Element => {
  const [inputValue, setInputValue] = useState("");
  const [editing, setEditing] = useState(false);
  function onKeyDown(event: React.KeyboardEvent<HTMLInputElement>): void {
    if (event.key === "Enter") {
      writePv(
        props.pvName,
        new DType({ stringValue: event.currentTarget.value })
      );
      setInputValue("");
      setEditing(false);
      event.currentTarget.blur();
    }
  }
  function onChange(event: React.ChangeEvent<HTMLInputElement>): void {
    setInputValue(event.currentTarget.value);
  }
  function onClick(event: React.MouseEvent<HTMLInputElement>): void {
    /* When focus gained allow editing. */
    if (!props.readonly && !editing) {
      setInputValue("");
      setEditing(true);
    }
  }
  function onBlur(event: React.ChangeEvent<HTMLInputElement>): void {
    setEditing(false);
    /* When focus lost show PV value. */
    setInputValue(DType.coerceString(props.value));
  }

  if (!editing && inputValue !== DType.coerceString(props.value)) {
    setInputValue(DType.coerceString(props.value));
  }

  const alarmQuality = props.value?.getAlarm().quality ?? AlarmQuality.VALID;

  return (
    <InputComponent
      pvName={props.pvName}
      value={inputValue}
      alarm={alarmQuality}
      alarmSensitive={props.alarmSensitive || false}
      readonly={props.readonly}
      foregroundColor={props.foregroundColor}
      backgroundColor={props.backgroundColor}
      transparent={props.transparent ?? false}
      onKeyDown={onKeyDown}
      onChange={onChange}
      onBlur={onBlur}
      onClick={onClick}
      font={props.font}
      textAlign={props.textAlign}
    />
  );
};

const InputWidgetProps = {
  ...PVWidgetPropType,
  font: FontPropOpt,
  foregroundColor: ColorPropOpt,
  backgroundColor: ColorPropOpt,
  transparent: BoolPropOpt,
  alarmSensitive: BoolPropOpt,
  textAlign: ChoicePropOpt(["left", "center", "right"])
};

export const Input = (
  props: InferWidgetProps<typeof InputWidgetProps>
): JSX.Element => <Widget baseWidget={SmartInputComponent} {...props} />;

registerWidget(Input, InputWidgetProps, "input");
