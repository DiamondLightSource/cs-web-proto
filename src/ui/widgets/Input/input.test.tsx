import React from "react";
import { SmartInputComponent } from "./input";
import { render } from "@testing-library/react";
import { DAlarm } from "../../../types/dtypes";
import { dstring } from "../../../testResources";

let input: JSX.Element;

beforeEach((): void => {
  input = (
    <SmartInputComponent
      pvName="pv"
      value={dstring("hello", DAlarm.MINOR)}
      connected={true}
      readonly={true}
      alarmSensitive={true}
    />
  );
});
describe("<Input />", (): void => {
  it("renders an input", (): void => {
    const { getByDisplayValue } = render(input);
    const renderedInput = getByDisplayValue("hello");
    expect(renderedInput).toBeInTheDocument();
    expect(renderedInput).toHaveClass("warning");
    expect(renderedInput).toHaveClass("readonly");
  });
});
