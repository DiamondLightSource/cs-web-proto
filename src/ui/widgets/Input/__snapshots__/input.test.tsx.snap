// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input Input /> it matches the snapshot 1`] = `
<input
  className="Input Input"
  onBlur={[Function]}
  onChange={[Function]}
  onClick={[Function]}
  onKeyDown={[Function]}
  readOnly={false}
  style={
    Object {
      "backgroundColor": undefined,
      "color": undefined,
    }
  }
  type="text"
  value="hello"
/>
`;
